spec.html

```diff
@@ -7628,19 +7627,44 @@ <h1>Jobs and Host Operations to Enqueue Jobs</h1>
+      <p>The specific choice of Realm is up to the host environment. This initial execution context and Realm is only in use before any callback function is invoked. When a callback function related to a Job, like a Promise handler, is invoked, the invocation pushes its own execution context and Realm.</p>
+    </emu-note>
+
+    <p>Particular kinds of Jobs have additional conformance requirements.</p>
+
     <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
       <h1>HostEnqueuePromiseJob ( _job_, _realm_ )</h1>
       <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the Job Abstract Closure _job_ to be performed, at some future time. The Abstract Closures used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
-      <p>The _realm_ parameter is passed through to hosts with no normative requirements; it is either *null* or a Realm.</p>
+      <p>The _realm_ parameter is either *null* or a Realm Record.</p>
+
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>. Additionally, Jobs enqueued by this hook must conform to all of the following requirements:</p>
+      <ul>
+        <li>If _realm_ is not *null*, the implementation must perform implementation-defined steps such that _job_ is prepared to evaluate code.</li>
+        <li>Let _scriptOrModule_ be GetActiveScriptOrModule(). The implementation must perform implementation-defined steps such that _job_ propagates the script or module _scriptOrModule_.</li>
+        <li>Jobs must be scheduled in FIFO order, with Jobs running in the same order as the HostEnqueuePromiseJob invocations which scheduled them.</li>
+      </ul>
 
       <emu-note>
-        The _realm_ for PromiseResolveThenableJobs is the result of calling GetFunctionRealm on the _then_ function object. The _realm_ for PromiseReactionJobs is the result of calling GetFunctionRealm on the handler if the handler is not *undefined*. Otherwise the _realm_ is *null*. The WHATWG HTML specification, for example, uses _realm_ to check for ability to run script and to prepare to run script.
+        <p>The _realm_ for PromiseResolveThenableJobs is usually the result of calling GetFunctionRealm on the _then_ function object. The _realm_ for PromiseReactionJobs is usually the result of calling GetFunctionRealm on the handler if the handler is not *undefined*. When the _realm_ is *null* due to being called on a revoked Proxy, the current Realm at the time of the GetFunctionRealm call is used instead. When the _realm_ is *null* and no user ECMAScript code will be evaluated, and no new ECMAScript objects (e.g. Error objects) will be created. The WHATWG HTML specification, for example, uses _realm_ to check for ability to run script and for the <a href="https://html.spec.whatwg.org/#entry">entry</a> concept.</p>
```

Comment:
I'm not actually sure which function "the handler" refers to here. It's obvious when the Job is produced by `NewPromiseReactionJob`, which does explicitly talk about a handler, but not when it is produced by `NewPromiseResolveThenableJob`, which does not. 

> `PromiseResolveThenableJobs`

This isn't actually a thing. (That's a bug in master, too.)

> When the _realm_ is *null* and no user ECMAScript code will be evaluated, and no new ECMAScript objects (e.g. Error objects) will be created.

This fragment lacks a verb; it is not a sentence.

Reply:
> I'm not actually sure which function "the handler" refers to here. 

The sentence about PromiseResolveThenableJob talks about the _then_ function, not the handler. The handler only refers to the reaction jobs.

> This isn't actually a thing. (That's a bug in master, too.)

Will fix.

> This fragment lacks a verb; it is not a sentence.

Oops, good catch.